File: C#/p001.cs
/*
    Puzzle 001 - Find programming errors and the SQL-inject vulnerability
*/

using System.Data.SqlClient;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;

namespace CSharp.Controllers
{
    [Route("api/sql")]
    public class SqlInjectionController : Controller
    {

        private readonly IConfiguration _configuration;
        public SqlInjectionController(IConfiguration configuration)
        {
            _configuration = configuration;
        }

        [HttpGet]
        [Route("user")]
        public string GetDataById()
        {
            using (SqlConnection connection = new SqlConnection(_configuration.GetValue<string>("ConnectionString")))
            {
                connection.Open();
                SqlCommand command = new SqlCommand("SELECT * FROM Users WHERE Id = "+ Request.Query["id"], connection);
                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        string returnString = string.Empty;
                        returnString += "Name :" + reader["Name"] + ". ";
                        returnString += "Description : " + reader["Description"];
                        return returnString;
                    }
                    else
                    {
                        return string.Empty;
                    }
                }
            }
        }
    }
}

File: Node/p001.js
const express = require('express');
const mysql = require('mysql2');
const config = require('./config'); // Assuming configuration is stored in a separate file

const app = express();

// Create a MySQL connection
const connection = mysql.createConnection({
  host: config.host,
  user: config.user,
  password: config.password,
  database: config.database
});

app.get('/api/sql/user', (req, res) => {
  const id = req.query.id;
  
  const query = `SELECT * FROM Users WHERE Id = ${id}`;
  
  connection.query(query, (error, results) => {
    if (error) {
      return res.status(500).json({ error: 'Database error' });
    }
    
    if (results.length > 0) {
      const user = results[0];
      const returnString = `Name: ${user.Name}. Description: ${user.Description}`;
      res.send(returnString);
    } else {
      res.send('');
    }
  });
});

const port = 3000;
app.listen(port, () => {
  console.log(`Server running on port ${port}`);
});